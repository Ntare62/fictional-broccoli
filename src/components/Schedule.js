import { useState, useEffect } from "react";

import { makeStyles } from "@material-ui/core/styles";

// import MUI list components
import ListSubheader from "@material-ui/core/ListSubheader";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItemText from "@material-ui/core/ListItemText";
import Collapse from "@material-ui/core/Collapse";
import ExpandLess from "@material-ui/icons/ExpandLess";
import ExpandMore from "@material-ui/icons/ExpandMore";

// import icons to work with within the schedule
import ViewDayIcon from "@material-ui/icons/ViewDay";
import QueueMusic from "@material-ui/icons/QueueMusic";

export default function Schedule() {
  const [shows, setShows] = useState();
  const [daysCollapsed, setDaysCollapsed] = useState(Array(7).fill(true));

  const days = [
    { name: "Sunday", num: 0 },
    { name: "Monday", num: 1 },
    { name: "Tuesday", num: 2 },
    { name: "Wednesday", num: 3 },
    { name: "Thursday", num: 4 },
    { name: "Friday", num: 5 },
    { name: "Saturday", num: 6 },
  ];

  // get show data from server
  useEffect(() => {
    const getShows = async () => {
      const response = await fetch("/api/shows/");
      if (!response.ok) {
        throw new Error(response.statusText);
      }
      const showResults = await response.json();
      setShows(showResults);
    };
    getShows();
  }, []);

  const useStyles = makeStyles((theme) => ({
    root: {
      width: "100%",
      minWidth: 400,
    },
    nested: {
      paddingLeft: theme.spacing(4),
    },
  }));

  const classes = useStyles();

  const changeDayCollapsed = (dayNum) => {
    const newDaysCollapsed = [...daysCollapsed];
    newDaysCollapsed[dayNum] = !newDaysCollapsed[dayNum];
    setDaysCollapsed(newDaysCollapsed);
  };

  // list of shows for a given day
  const showList = (dayNum) => {
    // we add 1 because days are 1-7 in database
    const showObjs =
      shows &&
      shows
        .filter((show) => show.schedule.charAt(1) === (dayNum + 1).toString())
        .sort(
          (a, b) =>
            parseInt(a.schedule.slice(3, 5)) - parseInt(b.schedule.slice(3, 5))
        );

    return showObjs.map((showObj) => (
      <List key={showObj.id} component="div" disablePadding>
        <ListItem className={classes.nested}>
          <ListItemIcon>
            <QueueMusic />
          </ListItemIcon>
          <ListItemText
            primary={`${showObj.title} - ${showObj.dj_name}`}
            secondary={`${showObj.schedule.slice(
              3,
              8
            )} - ${showObj.schedule.slice(9)}`}
          />
        </ListItem>
      </List>
    ));
  };

  // list of days of the week
  const dayList = days.map((day) => {
    const isCollapsed = daysCollapsed[day.num];

    return (
      <div key={day.num}>
        <ListItem button key={day} onClick={() => changeDayCollapsed(day.num)}>
          <ListItemIcon>
            <ViewDayIcon />
          </ListItemIcon>
          <ListItemText primary={day.name} />
          {isCollapsed ? <ExpandMore /> : <ExpandLess />}
        </ListItem>

        <Collapse in={!isCollapsed} timeout="auto" unmountOnExit>
          {shows && showList(day.num)}
        </Collapse>
      </div>
    );
  });

  return (
    <div>
      <List
        component="nav"
        aria-labelledby="nested-list-subheader"
        subheader={
          <ListSubheader component="div" 
          disableSticky={true}
          id="nested-list-subheader">
            Schedule for the Week
          </ListSubheader>
        }
        className={classes.root}
      >
        {dayList}
      </List>
    </div>
  );
}
