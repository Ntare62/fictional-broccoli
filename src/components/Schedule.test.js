import { render, screen, fireEvent, act } from "@testing-library/react";
import fetchMock from "fetch-mock-jest";
import Schedule from "./Schedule";

const shows = require("../../data/showseed.json");

describe.only("Schedule tests", () => {
  let localShows;

  beforeAll(() => {
    localShows = shows.map((show) => ({ ...show }));

    fetchMock.reset();
    fetchMock.get(`/api/shows/`, () => localShows);
  });

  test("Snapshot test", async () => {
    const { container } = render(<Schedule />);
    await act(async () => {
      await fetchMock.flush(true);
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  test("Clicking day makes shows visible", async () => {
    render(<Schedule />);
    await act(async () => {
      await fetchMock.flush(true);
    });

    const timeString = /\d\d:\d\d - \d\d:\d\d/;

    let saturdayListItem = screen.queryByText("Saturday");
    const numberSaturdayShows = localShows.filter(
      (show) => show.schedule.charAt(1) === "7"
    ).length;

    let visibleShows = screen.queryAllByText(timeString);
    expect(saturdayListItem).toBeInTheDocument();
    expect(visibleShows.length).toBe(0);

    fireEvent.click(saturdayListItem);
    visibleShows = screen.queryAllByText(timeString);
    expect(visibleShows).toHaveLength = numberSaturdayShows;

    saturdayListItem = screen.queryByText("Saturday");
    fireEvent.click(saturdayListItem);
    visibleShows = screen.queryAllByText(timeString);

    // TODO visible shows retains old value after click
    // expect(visibleShows[0]).not.toBeVisible()
    // expect(visibleShows).toHaveLength(0);
  });

  test("title and DJ name are displayed", async () => {
    const { getByText } = await render(<Schedule />);

    await act(async () => {
      await fetchMock.flush(true);
    });

    // pick a show on Sunday
    const show = localShows.find((s) => s.schedule.charAt(1) === "1");

    // click the Sunday drop down
    const sundayListItem = screen.queryByText("Sunday");
    fireEvent.click(sundayListItem);

    const displayedShow = getByText(`${show.title} -${show.dj_name}`);

    expect(displayedShow).toBeInTheDocument();
    expect(displayedShow).toBeVisible();
  });
});
