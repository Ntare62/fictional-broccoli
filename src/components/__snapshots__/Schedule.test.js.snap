// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schedule tests Snapshot test 1`] = `
<div>
  <nav
    aria-labelledby="nested-list-subheader"
    class="MuiList-root makeStyles-root-1 MuiList-padding MuiList-subheader"
  >
    <div
      class="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
      id="nested-list-subheader"
    >
      Schedule for the Week
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Sunday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Monday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Tuesday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Wednesday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Thursday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Friday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div>
      <div
        aria-disabled="false"
        class="MuiButtonBase-root MuiListItem-root makeStyles-header-2 MuiListItem-gutters MuiListItem-button"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2 21h19v-3H2v3zM20 8H3c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1h17c.55 0 1-.45 1-1V9c0-.55-.45-1-1-1zM2 3v3h19V3H2z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Saturday
          </span>
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          data-testid="isNotExpanded"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
  </nav>
</div>
`;
